opm/autodiff/NewtonIterationBlackoilInterleaved.cpp

```diff
@@ -48,245 +56,456 @@ namespace Opm
+
+            // Copy solver output to dx.
+            for (int i = 0; i < size; ++i) {
+                for( int p=0, idx = i; p<np; ++p, idx += size ) {
+                    dx(idx) = x[i][p];
                 }
             }
+
+            if ( hasWells ) {
+                // Compute full solution using the eliminated equations.
+                // Recovery in inverse order of elimination.
+                dx = recoverVariable(elim_eqs[1], dx, np);
+                dx = recoverVariable(elim_eqs[0], dx, np);
+            }
+            return dx;
         }
+
+    protected:
+        mutable int iterations_;
+        boost::any parallelInformation_;
+
+        const bool newton_use_gmres_;
+        const double linear_solver_reduction_;
+        const int    linear_solver_maxiter_;
+        const int    linear_solver_restart_;
+        const int    linear_solver_verbosity_;
+
+    }; // end NewtonIterationBlackoilInterleavedImpl
+
+
+
+    /// Construct a system solver.
+    NewtonIterationBlackoilInterleaved::NewtonIterationBlackoilInterleaved(const parameter::ParameterGroup& param,
+                                                                           const boost::any& parallelInformation_arg)
+      : newtonIncrement_(),
+        param_( param ),
+        parallelInformation_(parallelInformation_arg),
+        iterations_( 0 )
+    {
     }
 
+    namespace detail {
+
+        template< int NP >
+        struct NewtonIncrement
+        {
+            template <class NewtonIncVector>
+            static const NewtonIterationBlackoilInterface&
+            get( NewtonIncVector& newtonIncrements,
+                 const parameter::ParameterGroup& param,
+                 const boost::any& parallelInformation,
+                 const int np )
+            {
+                if( np == NP )
+                {
+                    assert( np < int(newtonIncrements.size()) );
+                    // create NewtonIncrement with fixed np
+                    if( ! newtonIncrements[ NP ] )
+                        newtonIncrements[ NP ].reset( new NewtonIterationBlackoilInterleavedImpl< NP >( param, parallelInformation ) );
+                    return *(newtonIncrements[ NP ]);
+                }
+                else
+                {
+                    return NewtonIncrement< NP-1 >::get(newtonIncrements, param, parallelInformation, np );
+                }
+            }
+        };
 
+        template<>
+        struct NewtonIncrement< 1 >
+        {
+            template <class NewtonIncVector>
+            static const NewtonIterationBlackoilInterface&
+            get( NewtonIncVector& newtonIncrements,
+                 const parameter::ParameterGroup& param,
+                 const boost::any& parallelInformation,
+                 const int np )
+            {
+                OPM_THROW(std::runtime_error,"NewtonIncrement::get: number of variables not supported yet. Adjust maxNumberEquations appropriately to cover np = " << np);
```

Comment:
1. Would it not be better to use <0> as the terminating specialization? While slightly inefficient, the 1x1 block case is a special case that would work well.
2. Some of the arguments of this function are unused, giving me warnings.
3. Would not the return statement be a segfault (if the throw was not there)? Would it not be better to just leave it out?